### PR TITLE
Gumbel MoE SRF: per-node expert routing in surface decoder

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The SRF uses a single MLP for all surface nodes regardless of physical regime. Surface nodes fall into distinct regimes: boundary layer, wake, leading edge (stagnation + suction peak), trailing edge (pressure recovery). A **Mixture-of-Experts SRF with hard Gumbel routing** allows regime-specialized experts.

Different from: MoE FFN (#2268, routed FFN in backbone), MoE output routing (exhausted, routed outputs). This is hard per-NODE routing at the SRF hidden layer. Reference: Switch Transformer (Fedus et al. 2021) — hard top-1 outperforms soft routing when experts can meaningfully specialize.

### Changes to `cfd_tandemfoil/train.py`

1. **Add flags:**
   ```python
   parser.add_argument('--gumbel_moe_srf', action='store_true')
   parser.add_argument('--gumbel_k', type=int, default=4, help='Number of SRF experts')
   parser.add_argument('--gumbel_lb_weight', type=float, default=0.01, help='Load balancing loss weight')
   ```

2. **Implement GumbelMoESRF:**
   ```python
   class GumbelMoESRF(nn.Module):
       def __init__(self, d_input, d_hidden=192, d_output=3, K=4):
           super().__init__()
           self.K = K
           self.input_proj = nn.Linear(d_input, d_hidden)
           self.router = nn.Linear(d_hidden, K)
           self.experts = nn.ModuleList([
               nn.Sequential(nn.GELU(), nn.Linear(d_hidden, d_hidden), nn.GELU(), nn.Linear(d_hidden, d_hidden))
               for _ in range(K)
           ])
           self.output_proj = nn.Linear(d_hidden, d_output)
       
       def forward(self, x, tau=1.0):
           h = self.input_proj(x)
           logits = self.router(h.detach())  # detach prevents router affecting features
           
           if self.training:
               gumbel = -torch.log(-torch.log(torch.rand_like(logits) + 1e-8) + 1e-8)
               soft = F.softmax((logits + gumbel) / tau, dim=-1)
               hard = F.one_hot(soft.argmax(-1), self.K).float()
               routing = hard - soft.detach() + soft
           else:
               routing = F.one_hot(logits.argmax(-1), self.K).float()
           
           expert_out = torch.stack([e(h) for e in self.experts], dim=-2)
           h_expert = (expert_out * routing.unsqueeze(-1)).sum(-2)
           out = self.output_proj(h_expert)
           
           f = routing.mean(dim=[0,1])
           p = F.softmax(logits, dim=-1).mean(dim=[0,1])
           lb_loss = self.K * (f * p).sum()
           return out, lb_loss
   ```

3. **Temperature annealing:** `tau = max(0.1, 1.0 * (0.1 ** (epoch/max_epochs)))`

4. **Add lb_loss:** `total_loss += args.gumbel_lb_weight * lb_loss`

5. **Apply to BOTH fore-foil and aft-foil SRF** (separate instances).

6. **Log to W&B:** `train/moe_expert_usage`, `train/moe_lb_loss`, `train/moe_tau`

7. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent askeladd --wandb_name "askeladd/gumbel-moe-k4-s42" --wandb_group gumbel-moe-srf \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
     --gumbel_moe_srf --gumbel_k 4 --gumbel_lb_weight 0.01

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent askeladd --wandb_name "askeladd/gumbel-moe-k4-s73" --wandb_group gumbel-moe-srf \
     --seed 73 [same flags as above]
   ```

8. If expert collapse (>60% to one expert), try K=2 in follow-up.

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Gumbel MoE SRF (K=4 experts) — 2-Seed Results

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | Δ |
|--------|----------|---------|---------|------------|---|
| **p_in** | **11.709** | 12.2 | 11.9 | **12.05** | **+2.9% ❌** |
| **p_oodc** | **7.544** | 7.4 | 7.7 | **7.55** | **+0.1% ≈** |
| **p_tan** | **27.402** | 27.4 | 27.8 | **27.6** | **+0.7% ≈** |
| p_re | 6.481 | 6.6 | 6.2 | **6.4** | **-1.2% ✅** |

### Experiment Details

**Commands used:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/gumbel-moe-k4-s42" --wandb_group gumbel-moe-srf \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --gumbel_moe_srf --gumbel_k 4 --gumbel_lb_weight 0.01

# Same for seed 73 on GPU 1
```

**W&B runs:**
- Seed 42: `jmzq4sft` (best epoch 159, p_in=12.2, p_tan=27.4)
- Seed 73: `xgh8j9mz` (best epoch 158, p_in=11.9, p_tan=27.8)

**Peak memory:** ~75-93 GB
**Epochs completed:** 159/158 (more than baseline ~149 — MoE adds negligible overhead to surface-only computation)

### Implementation Notes

- `GumbelMoESRF` class replaces `SurfaceRefinementHead` when `--gumbel_moe_srf` is set
- K=4 experts, each is a 2-layer MLP (GELU-Linear-GELU-Linear), shared input_proj and output_proj
- Router input is detached from features (prevents router from affecting backbone representations)
- Gumbel-Softmax with straight-through estimator for hard top-1 routing
- Temperature annealing: tau = max(0.1, 1.0 × 0.1^(epoch/max_epochs))
- Load balancing loss (Switch Transformer style): K × Σ(f_i × p_i), weight=0.01
- Zero-init output for identity initialization
- All expert usage fractions logged to W&B (`train/moe_expert_*_frac`)

### What happened — honest analysis

**Neutral result.** The MoE SRF matches baseline on OOD metrics (p_oodc flat, p_re slightly improved) but regresses slightly on p_in (+2.9%). Not a clear win.

Key observations:
1. **No throughput penalty**: 159/158 epochs vs ~149 baseline. The MoE operates only on surface nodes, so the 4 separate expert forward passes add negligible overhead.
2. **p_re improvement** (-1.2%) is interesting — expert specialization might help with extreme-Re surface nodes — but within noise range.
3. **p_in regression** (+2.9%) suggests the routing + load balancing overhead hurts the model's ability to learn the in-distribution pattern compared to a single dedicated MLP.
4. The load balancing loss (weight=0.01) may be too aggressive, forcing equal expert utilization when unequal is physically appropriate (more nodes in boundary layer, fewer at LE/TE).

### Suggested follow-ups

- **Reduce lb_weight to 0.001**: Allow more imbalanced routing — let the model decide if 3 experts handle boundary layer while 1 handles LE.
- **K=2 instead of K=4**: Simpler routing might work better — fore-foil vs aft-foil or pressure-side vs suction-side.
- **NOT recommended for merge**: The +2.9% p_in regression doesn't justify the added complexity.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2348 because ADVISOR: Closing — Gumbel MoE SRF is approximately neutral (p_in +2.9%, p_oodc flat, p_tan +0.7%, p_re -1.2%). The per-node expert routing doesn't provide meaningful benefit over the single SRF MLP. The load balancing loss may force equal expert utilization when physically inappropriate. Good implementation though. MoE in the surface decoder is now exhausted..
